File: web/src/lib/faq-metadata.ts
export type FAQProtocol = "bisq_easy" | "multisig_v1" | "musig" | "all";

export const FAQ_PROTOCOL_OPTIONS: ReadonlyArray<{ value: FAQProtocol; label: string }> = [
  { value: "bisq_easy", label: "Bisq Easy" },
  { value: "multisig_v1", label: "Bisq 1 (Multisig)" },
  { value: "musig", label: "MuSig" },
  { value: "all", label: "All Protocols" },
];

export const FAQ_CATEGORIES: ReadonlyArray<string> = [
  "General",
  "Trading",
  "Wallet",
  "Security",
  "Reputation",
  "Payments",
  "Technical",
  "Bisq Easy",
  "Bisq 2",
  "Fees",
  "Account",
];

const CATEGORY_KEYWORDS: ReadonlyArray<{ category: string; tokens: string[]; weight: number }> = [
  {
    category: "Trading",
    tokens: ["trade", "offer", "maker", "taker", "price", "spread", "mediation", "arbitration", "settle"],
    weight: 3,
  },
  {
    category: "Wallet",
    tokens: ["wallet", "seed", "restore", "backup", "address", "txid", "transaction id", "utxo", "keys"],
    weight: 3,
  },
  {
    category: "Security",
    tokens: ["security", "scam", "phishing", "tor", "pgp", "signature", "verify", "fraud"],
    weight: 3,
  },
  {
    category: "Reputation",
    tokens: ["reputation", "profile age", "profile", "badge", "score", "burn bsq"],
    weight: 3,
  },
  {
    category: "Payments",
    tokens: ["payment", "bank", "sepa", "iban", "wise", "revolut", "zelle", "fiat", "ach"],
    weight: 3,
  },
  {
    category: "Technical",
    tokens: ["error", "crash", "install", "upgrade", "update", "version", "log", "broken", "stuck"],
    weight: 2,
  },
  {
    category: "Fees",
    tokens: ["fee", "fees", "mining fee", "network fee"],
    weight: 2,
  },
  {
    category: "Account",
    tokens: ["account", "login", "sign in", "identity", "session"],
    weight: 2,
  },
  {
    category: "Bisq Easy",
    tokens: ["bisq easy"],
    weight: 2,
  },
  {
    category: "Bisq 2",
    tokens: ["bisq 2", "bisq2"],
    weight: 2,
  },
];

function sanitizeText(value: string | null | undefined): string {
  return (value || "").toLowerCase();
}

export function detectFaqProtocol(...texts: Array<string | null | undefined>): FAQProtocol {
  const combined = texts.map(sanitizeText).join(" ");
  const hasBisq2 = combined.includes("bisq easy") || combined.includes("bisq 2") || combined.includes("bisq2");
  const hasBisq1 = combined.includes("bisq 1") || combined.includes("bisq1") || combined.includes("multisig");
  const hasMuSig = combined.includes("musig");

  if (hasMuSig && !hasBisq1 && !hasBisq2) return "musig";
  if (hasBisq1 && hasBisq2) return "all";
  if (hasBisq1) return "multisig_v1";
  if (hasBisq2) return "bisq_easy";
  return "all";
}

export function detectFaqCategory(
  question: string | null | undefined,
  answer?: string | null | undefined,
  fallbackCategory?: string | null | undefined,
): string {
  const preferred = (fallbackCategory || "").trim();
  if (preferred && preferred.toLowerCase() !== "general") return preferred;

  const combined = `${sanitizeText(question)} ${sanitizeText(answer)}`.trim();
  if (!combined) return "General";

  const scores: Record<string, number> = {};
  for (const category of FAQ_CATEGORIES) scores[category] = 0;

  for (const entry of CATEGORY_KEYWORDS) {
    for (const token of entry.tokens) {
      if (combined.includes(token)) {
        scores[entry.category] += entry.weight;
      }
    }
  }

  let bestCategory = "General";
  let bestScore = 0;
  for (const [category, score] of Object.entries(scores)) {
    if (score > bestScore) {
      bestCategory = category;
      bestScore = score;
    }
  }

  if (bestScore > 0) return bestCategory;
  return preferred || "General";
}

export function inferFaqMetadata(params: {
  question?: string | null;
  answer?: string | null;
  category?: string | null;
  protocol?: FAQProtocol | null;
}): { category: string; protocol: FAQProtocol } {
  const category = detectFaqCategory(params.question, params.answer, params.category);
  const protocol = params.protocol || detectFaqProtocol(params.question, params.answer);
  return { category, protocol };
}

File: api/app/services/faq/faq_metadata.py
"""FAQ metadata normalization and inference helpers.

These helpers ensure new FAQs always have both `category` and `protocol` set.
"""

from typing import Dict, Literal, Optional, Tuple, cast

FAQProtocol = Literal["multisig_v1", "bisq_easy", "musig", "all"]

_CATEGORY_KEYWORDS = {
    "Trading": (
        3,
        [
            "trade",
            "offer",
            "maker",
            "taker",
            "price",
            "spread",
            "mediation",
            "arbitration",
            "settle",
        ],
    ),
    "Wallet": (
        3,
        [
            "wallet",
            "seed",
            "restore",
            "backup",
            "address",
            "txid",
            "transaction id",
            "utxo",
            "keys",
        ],
    ),
    "Security": (
        3,
        ["security", "scam", "phishing", "tor", "pgp", "signature", "verify", "fraud"],
    ),
    "Reputation": (
        3,
        ["reputation", "profile age", "profile", "badge", "score", "burn bsq"],
    ),
    "Payments": (
        3,
        ["payment", "bank", "sepa", "iban", "wise", "revolut", "zelle", "fiat", "ach"],
    ),
    "Technical": (
        2,
        [
            "error",
            "crash",
            "install",
            "upgrade",
            "update",
            "version",
            "log",
            "broken",
            "stuck",
        ],
    ),
    "Fees": (2, ["fee", "fees", "mining fee", "network fee"]),
    "Account": (2, ["account", "login", "sign in", "identity", "session"]),
    "Bisq Easy": (2, ["bisq easy"]),
    "Bisq 2": (2, ["bisq 2", "bisq2"]),
}


def _sanitize(text: Optional[str]) -> str:
    return (text or "").lower()


def infer_protocol(question: Optional[str], answer: Optional[str]) -> FAQProtocol:
    """Infer FAQ protocol from question/answer content."""
    combined = f"{_sanitize(question)} {_sanitize(answer)}".strip()
    has_bisq2 = "bisq easy" in combined or "bisq 2" in combined or "bisq2" in combined
    has_bisq1 = "bisq 1" in combined or "bisq1" in combined or "multisig" in combined
    has_musig = "musig" in combined

    if has_musig and not has_bisq1 and not has_bisq2:
        return "musig"
    if has_bisq1 and has_bisq2:
        return "all"
    if has_bisq1:
        return "multisig_v1"
    if has_bisq2:
        return "bisq_easy"
    return "all"


def infer_category(
    question: Optional[str],
    answer: Optional[str],
    fallback_category: Optional[str] = None,
) -> str:
    """Infer FAQ category from content, preserving explicit user choice."""
    preferred = (fallback_category or "").strip()
    if preferred and preferred.lower() != "general":
        return preferred

    combined = f"{_sanitize(question)} {_sanitize(answer)}".strip()
    if not combined:
        return "General"

    scores: Dict[str, int] = {category: 0 for category in _CATEGORY_KEYWORDS}
    for category, (weight, keywords) in _CATEGORY_KEYWORDS.items():
        for token in keywords:
            if token in combined:
                scores[category] += weight

    best_category = "General"
    best_score = 0
    for category, score in scores.items():
        if score > best_score:
            best_category = category
            best_score = score

    if best_score > 0:
        return best_category
    return preferred or "General"


def normalize_faq_metadata(
    question: Optional[str],
    answer: Optional[str],
    category: Optional[str],
    protocol: Optional[str],
) -> Tuple[str, FAQProtocol]:
    """Return normalized (category, protocol) for FAQ creation."""
    normalized_category = infer_category(question, answer, category)
    if protocol in {"multisig_v1", "bisq_easy", "musig", "all"}:
        normalized_protocol = cast(FAQProtocol, protocol)
    else:
        normalized_protocol = infer_protocol(question, answer)
    return normalized_category, normalized_protocol
